Tests/UnitTests/Core/EventData/MeasurementTests.cpp

```diff
@@ -1,86 +1,236 @@
 // This file is part of the Acts project.
 //
-// Copyright (C) 2016-2018 CERN for the benefit of the Acts project
+// Copyright (C) 2016-2020 CERN for the benefit of the Acts project
 //
 // This Source Code Form is subject to the terms of the Mozilla Public
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
+#include <boost/test/data/test_case.hpp>
 #include <boost/test/unit_test.hpp>
 
 #include "Acts/EventData/Measurement.hpp"
-#include "Acts/EventData/MeasurementHelpers.hpp"
-#include "Acts/Surfaces/CylinderSurface.hpp"
+#include "Acts/Tests/CommonHelpers/FloatComparisons.hpp"
+#include "Acts/Tests/CommonHelpers/GenerateParameters.hpp"
 #include "Acts/Tests/CommonHelpers/TestSourceLink.hpp"
-#include "Acts/Utilities/ParameterDefinitions.hpp"
 
+#include <limits>
 #include <random>
+#include <tuple>
+#include <vector>
+
+using namespace Acts;
+using namespace Acts::Test;
+using SourceLink = Acts::Test::TestSourceLink;
+namespace bd = boost::unit_test::data;
+
+namespace {
+constexpr BoundIndices boundIndices[] = {
+    eBoundLoc0, eBoundLoc1, eBoundTime, eBoundPhi, eBoundTheta, eBoundQOverP,
+};
+constexpr FreeIndices freeIndices[] = {
+    eFreePos0, eFreePos1, eFreePos2, eFreeTime,
+    eFreeDir0, eFreeDir1, eFreeDir2, eFreeQOverP,
+};
+const TestSourceLink source;
+// fix seed for reproducible tests
+std::default_random_engine rng(123);
+}  // namespace
+
+// the underlying subspace implementation is already tested exhaustively in a
+// separate unit test. here we only test concrete extreme cases and
+// measurement-specific functionality.
+
+BOOST_AUTO_TEST_SUITE(EventDataMeasurement)
+
+BOOST_DATA_TEST_CASE(FixedBoundOne, bd::make(boundIndices), index) {
+  auto [params, cov] = generateParametersCovariance(rng, index);
+  auto meas = makeMeasurement(source, params, cov, index);
+
+  BOOST_CHECK_EQUAL(meas.size(), 1);
+  for (auto i : boundIndices) {
+    if (i == index) {
+      BOOST_CHECK(meas.contains(i));
+    } else {
+      BOOST_CHECK(not meas.contains(i));
+    }
+  }
+  BOOST_CHECK_EQUAL(meas.parameters(), params);
+  BOOST_CHECK_EQUAL(meas.covariance(), cov);
+  BOOST_CHECK_EQUAL(meas.sourceLink(), source);
+}
+
+BOOST_AUTO_TEST_CASE(FixedBoundAll) {
+  auto [params, cov] = generateBoundParametersCovariance(rng);
+  auto meas = makeMeasurement(source, params, cov, eBoundLoc0, eBoundLoc1,
```

Comment:
I think I know why you're doing `auto` in so many places, and I won't be picky about this. But I think we should try to keep `auto` usage somewhat consistent. There is AAA, but so far we haven't followed it.

Reply:
What do you mean by AAA?

Reply:
Almost Always Auto, I guess.

Some people think that `auto` should be used wherever possible in C++, with the possible exception of function signatures (accidentally breaking API by changing an implementation is never nice), because it un-clutters code from type annotations which are often obvious.

Although I use this discipline in other languages with type inference like Rust, I would personally disagree with using it in C++, because unconditionally applying type inference without second thought puts you at the risk of accidentally changing semantics in a programming language with implicit conversions, constructors with assignment syntax, and constructor/assignment overloading. [See this Eigen documentation page for example](http://eigen.tuxfamily.org/dox/TopicPitfalls.html).

But given that simple policies like "always use auto" or "never use auto" fail in C++, it's hard to specify when auto should or should not be used. Personally, I tend to go for the minimal "if the reviewer thinks it's okay, it's okay" policy, and use auto only to avoid repetitions myself (e.g. when casting types, the type is already written in the cast).

Reply:
Thanks for the explanation. I agree with @HadrienG2 that neither of the extreme policies would work well for use. In the particular case of the `makeMeasurement` helper function, it is intended explicitly w/ `auto` in mind: by using automatic template type deduction in the function call we avoid having to explicitly define all template parameters of the `FixedSizedMeasurement` class.

Reply:
Ah yeah sorry for not going into detail there. Like I said, I won't be picky here, I just saw a large number of `auto`s in the PR and wanted to raise the point.

Reply:
I understand, I just wanted to clarify why I needed to be closer to the `always use auto` in this particular case.